### ISSUE TITLE
Weather applet for Budgie Desktop

### ISSUE THREAD

**Issue by Codalunga** [solus-project/budgie-desktop#689]
Operating System: e.g. Solus 1.2.1.
Budgie version: Budgie 10.2.8.

Maybe it's just me, but I'm used to check the weather with just a glimpse at the up-right corner of my screen, so a weather applet that can be easily added via Raven is sorely missed :D I don't think people are craving for this, but it's still be cool to have such a feature!

 ...

**Comment by tristan957** [comment]
I'd love a little applet in Raven to show current temp and stuff. An implementation like macOS' would be quite nice as an optional applet.

**Comment by ikeydoherty** [comment]
Screenie?

**Comment by tristan957** [comment]
![20161027_214001](https://cloud.githubusercontent.com/assets/11936772/19792870/17fda270-9c8f-11e6-8377-5c8125ff611f.jpg)

@ikeydoherty took this right off my friend's laptop. I think it has potential regarding a basic design. Would love to see how you could improve it.

**Comment by ikeydoherty** [comment]
I should murder you for the pissed screenshot. lol.

**Comment by Justinzobel** [comment]
A start on one was made by this user and code can be retrieved from there if anyone is interested http://entornosgnulinux.com/2016/05/21/minimalist-weather-indicator-en-solus-1-1/

**Comment by tristan957** [comment]
I just got back from basketball shenanigans. Cut me some slack Jack ;)

**Comment by ikeydoherty** [comment]
=P

**Comment by Codalunga** [comment]
Yeah I was just thinking something simple, similar to Ubuntu Weather Indicator, or any xfce/mate weather plugin :D You know, something similar to this
![weather](https://cloud.githubusercontent.com/assets/22096160/19798037/0033d32a-9cef-11e6-92b0-9e32c2e457a8.png)

@tristan957 that looks pretty rad, man. I can't figure out a way to implement it without adding another tab to Raven tough, which would be sort of redundant just for weather notifications, at least for me :D
@Justinzobel thanks for the heads-up, it needs more polish but it's promising!

**Comment by Vistaus** [comment]
I want such an applet as well! I've tried to install and setup the aforementioned Weather applet but it doesn't display the current temperature :cry: A real applet is what we need!

**Comment by ikeydoherty** [comment]
We support out of tree applets, there is literally nothing stopping anyone from writing one :) It's encouraged, I don't want bloat in the core of Budgie.

**Comment by tristan957** [comment]
How can we go about learning how to create an applet? I'd like to try my hand at programming that isn't related to schoolwork.

**Comment by ikeydoherty** [comment]
Good question! So, what resources can I help create, and what would be the best format?

**Comment by tristan957** [comment]
That's an even better question. I feel like there could be a video series or an online tutorial written out with screenshots that followed the development of an applet, which included commented source code to look at. Obviously I'm no expert, but maybe a page on the Wiki that was titled "For Developers" and had different headers or sub-pages like "Applet Development", "Budgie Development", "Packaging" (maybe packaging could be moved onto this page as well), etc. For example in "Applet Development", there could be "Creating the 'Hello World' applet", "How to integrate an applet into the panel", "How to integrate an applet into Raven", etc. There could also be opinions like what the preferred language is for creating an applet. Also helpful links to the GTK documentation as well as others. 

I think this would be a better suited question on Google+ and Reddit where you can get more opinions though.

**Comment by ikeydoherty** [comment]
Well asking on G+ is nice *but* leads to a lot of bikeshedding, where there isn't *genuine* interest in developing a Budgie applet, rather just a large box for people to shout into :)

**Comment by tristan957** [comment]
Maybe asking the IRC would be better. There are normally technologically literate people there :P

**Comment by ikeydoherty** [comment]
No :P **Summons IRC People**

.... **waits awkwardly** ...

**Comment by baimafeima** [comment]
> (...) but maybe a page on the Wiki that was titled "For Developers" and had different headers or sub-pages like "Applet Development", "Budgie Development", "Packaging" (maybe packaging could be moved onto this page as well), etc. 

@tristan957 I think this would be a good start.

**Comment by kyrios123** [comment]
I also hope that someone will develop such an applet: I'm using the MATE edition because my wife wants to see the weather in the panel  :sunny:  :partly_sunny:  :cloud:

**Comment by jawz101** [comment]
Now it's approved?  lol @kyrios123 good work, man.

**Comment by kyrios123** [comment]
lol @jawz101 
Yeah I am such an influential person, you know !  :trollface:

**Comment by ilgarmehmetali** [comment]
Is anyone working on this? im thinking of giving it a go.
I tried looking for a completely free to use weather api but couldn't find one. Would asking users to register a apikey from an online provider be okey?

**Comment by Vistaus** [comment]
@ilgarmehmetali 

You can use darksky.net's API. The API calls are limited to 1000 free calls a day (you can have more if you pay) so it's no ideal, but for a first applet, you should be okay with 1000 free calls. A subscription allows more calls, it costs money, but it's cheap and maybe we could work out some sort of crowdfunding. I love Dark Sky esp. for the fact that they're very accurate.

If you're not satisfied with that, you can also use the IMHO almost-as-good OpenWeatherMap, but everyone needs to register their own API so that's slightly inconvenient. But it's free, no limited API calls and having your own key as a user also has it's pro's :)

**Comment by tristan957** [comment]
Why not just use `libgweather` I think is the name of the library?

**Comment by ilgarmehmetali** [comment]
@Vistaus 
We would have to crowdfund it somehow if we are to use darksky;
> [**Can I require my users to sign up for a developer's API key?**](https://darksky.net/dev/docs/faq#sublicensing)
No, your end-users should not sign up for Dark Sky API keys as a means for you to circumvent the daily cap of 1,000 free calls. The 1,000 free calls per day are intended for one developer or company to explore the API.

@tristan957 
libgweather doesnt have all the cities in my country and they have 53 open bug reports about locations thats just sitting there. So i dont realy want to use it.

 I think i will use OpenWeatherMap and ask users to get apikey.

**Comment by tristan957** [comment]
I mean it doesn't seem so hard to contribute to libgweather. Why not contribute and make the library better? It says it only allows for cities with greater than 100,000 people. Shouldn't be too hard for your country.

**Comment by ilgarmehmetali** [comment]
Unfortunately there are no public metar stations in my city to contribute data for but in a later time i might try finding metar stations for other major cities in my country.
I think it would be hard to cover whole of the world in libgweather with people - companies etc selling their metar reports. Unfortunately not every weather station is sharing their data freely.

**Comment by ilgarmehmetali** [comment]
started working on this
https://github.com/ilgarmehmetali/budgie-weather-applet
its little rough on the edges but its working, i think :)

**Comment by Justinzobel** [comment]
Asking users to get an API key is just ugly :/ but I'll try out your implementation @ilgarmehmetali

**Comment by Vistaus** [comment]
@Justinzobel It's the same way the popular (!) OpenWeather extension for GNOME does it and no one complains about that. I don't think it's ugly, but it's not as user-friendly, of course. But hey, give the guy some credit. At least *he* cared enough to create a weather applet for us! And keep in mind that this is a first start, who knows, maybe later on he can rework it so that users don't need their own API key anymore.

@ilgarmehmetali Thank you so, so much for doing this! I'll try it out today :)

**Comment by Justinzobel** [comment]
@Vistaus you mistake my comments for insult, they are not. It's just that with Solus we try to make things as simple as possible for the end user.

**Comment by ilgarmehmetali** [comment]
I have added libgweather as a provider too. Things are working better now, check it out again.
@Justinzobel It seems OpenWeatherMap [supports foss](http://openweathermap.org/faq), i have requested them to remove limitations from my key too, they said they will consider it. Lets hope they decide to support my project too.

**Comment by Justinzobel** [comment]
Oh that's excellent news! I'll give it a try now.

**Comment by Justinzobel** [comment]
Great work already! Are you planning for it to be clickable to see forecast type information?

**Comment by ilgarmehmetali** [comment]
Yes, im thinking of adding forecast, but i cant give any estimated time for that.
Though i could say most of the work is already done, adding forecast shouldn't be too hard.
